quinn-proto/src/endpoint.rs

```diff
@@ -602,8 +607,29 @@ impl Endpoint {
         Ok(())
     }
 
+    /// Reject this incoming connection attempt
+    pub fn reject(&mut self, incoming: Incoming, buf: &mut BytesMut) -> Transmit {
+        self.initial_close(
+            incoming.version,
+            incoming.addresses,
+            &incoming.crypto,
+            &incoming.src_cid,
+            TransportError::CONNECTION_REFUSED(""),
+            buf,
+        )
+    }
+
     /// Respond with a retry packet, requiring the client to retry with address validation
-    fn retry(&mut self, incoming: Incoming, buf: &mut BytesMut) -> Transmit {
+    ///
+    /// Errors if `incoming.remote_address_validated()` is true.
+    pub fn retry(
+        &mut self,
+        incoming: Incoming,
+        buf: &mut BytesMut,
+    ) -> Result<Transmit, RetryError> {
+        if incoming.remote_address_validated() {
```

Comment:
This feels a little silly. Should `Incoming` be an `enum` instead with variants that signal whether the remote address has been validated, so that we make it statically impossible to trigger this error?

Reply:
Hm. While I like the impulse, I can't think of a good way to make it work.

The first problem is that there would be methods duplicated between the two specific types. We'd want to implement `accept`, `reject`, `ignore`, `local_ip`, and `remote_address` on both the retryable and non-retryable variants, and possibly the parent enum too. Possibly we could factor this out into a trait which all three implement, but overall that seems over-complicated.

A second problem would come up if we start sending `NEW_TOKEN` frames. That would introduce the possibility of an `Incoming` which is both validated and retryable. We would want our API to allow those to be retried, but a typical user application would not actually want to retry them. This would create an awkward situation with any enum-based way of structuring this I can think of.

As one example, if we established this as:

```rs
enum Incoming {
    Validated(ValidatedIncoming),
    NotValidated(NotValidatedIncoming),
}
```

Then introducing `NEW_TOKEN` frames would mean we would want to make `NotValidatedIncoming` sometimes retryable, so we'd be back to the error not being statically prevented.

As an alternative example, if we established this as:

```rs
enum Incoming {
    Retryable(RetryableIncoming),
    NotRetryable(NotRetryableIncoming),
}
```

That would solve the above issue but make applications implementing retry logic awkwardly complicated, because they'd want to be doing:

```rs
match incoming {
    Incoming::Retryable(incoming) => if incoming.remote_address_validated() {
        handle(incoming.accept())
    } else {
        incoming.retry()
    },
    Incoming::NotRetryable(incoming) => handle(incoming.accept()),
}
```

Which seems gross.

Having there just be a `remote_address_validated` method is nice in that it avoids those problems. If we introduce `NEW_TOKEN` frame sending then we can just also introduce a `may_retry` method, such that `!remote_address_validated` implies `may_retry` but not necessarily the other way around (like I originally had in my prototype [[link]](https://github.com/gretchenfrage/quinn/pull/1#issuecomment-1909192157)). Which would be nice because we'd be exposing maximal freedom but without requiring existing applications to contort themselves to adapt to it.

As to the awkwardness of this solution not statically preventing the error, I might compare that to say, how `std::fs::File` still implements `Write` even if you opened it in read-only mode.

Reply:
Thanks for the in-depth response!

> The first problem is that there would be methods duplicated between the two specific types. We'd want to implement `accept`, `reject`, `ignore`, `local_ip`, and `remote_address` on both the retryable and non-retryable variants, and possibly the parent enum too. Possibly we could factor this out into a trait which all three implement, but overall that seems over-complicated.

This doesn't seem like a big deal? We could use a private type that underlies both of the enum variant types and expose these methods only on the enum type.

> A second problem would come up if we start sending `NEW_TOKEN` frames. That would introduce the possibility of an `Incoming` which is both validated and retryable. We would want our API to allow those to be retried, but a typical user application would not actually want to retry them. This would create an awkward situation with any enum-based way of structuring this I can think of.

Not sure if this is true? 9000 says in 8.1.2:

> In response to processing an Initial packet containing a token that was provided in a Retry packet, a server cannot send another Retry packet; it can only refuse the connection or permit it to proceed.

So, if the client sends a token, the server is not allowed to send a retry. I'm not sure how you think this ties in with `NEW_TOKEN`, but from my skimming of the RFC it seems like `NEW_TOKEN` would never be sent in Initial packets, only on established connections -- but I can't find an authoritative reference for this.

Reply:
> Not sure if this is true?

@gretchenfrage's analysis is correct. `NEW_TOKEN` frames are sent by the server to provide a client with tokens that the client may then choose to supply in future Initial packets. Per your quote, because these tokens were not provided in a Retry packet, the server may send a Retry in response, even though a token was present.

Reply:
Ah, alright then! I retract my objection.